### PR TITLE
bpo-46906: Add PyFloat_Pack8() to the C API

### PR DESCRIPTION
**Pull request by vstinner**
Add new functions to pack and unpack C double (serialize and
deserialize):

* PyFloat_Pack2()
* PyFloat_Pack4()
* PyFloat_Pack8()
* PyFloat_Unpack2()
* PyFloat_Unpack4()
* PyFloat_Unpack8()

Document these functions and add unit tests.

Rename functions and move them from the internal C API to the public
C API:

* _PyFloat_Pack2() => PyFloat_Pack2()
* _PyFloat_Pack4() => PyFloat_Pack4()
* _PyFloat_Pack8() => PyFloat_Pack8()
* _PyFloat_Unpack2() => PyFloat_Unpack2()
* _PyFloat_Unpack4() => PyFloat_Unpack4()
* _PyFloat_Unpack8() => PyFloat_Unpack8()

Replace the "unsigned char*" type with "char*" which is more common
and easy to use.

https://bugs.python.org/issue46906

### CODE REVIEW

**Comment by vstinner** [comment]
cc @methane @mdickinson

File: Doc/c-api/float.rst [hunk 1]
```diff
@@ -76,3 +76,83 @@ Floating Point Objects
...
+and NaNs (if such things exist on the platform) isn't handled correctly, and
+attempting to unpack a bytes string containing an IEEE INF or NaN will raise an
+exception.
+
+On non-IEEE platforms with more precision, or larger dynamic range, than IEEE
+754 supports, not all values can be packed; on non-IEEE platforms with less
+precision, or smaller dynamic range, not all values can be unpacked. What
+happens in such cases is partly accidental (alas).
+
+.. verionadded:: 3.11
```
Comment (kumaraditya303) [inline Doc/c-api/float.rst]:
```suggestion
.. versionadded:: 3.11
```

File: Modules/_ctypes/cfield.c [hunk 2]
```diff
@@ -1057,10 +1056,10 @@ f_set_sw(void *ptr, PyObject *value, Py_ssize_t size)
     if (x == -1 && PyErr_Occurred())
         return NULL;
 #ifdef WORDS_BIGENDIAN
```
Comment (arhadthedev) [inline Modules/_ctypes/cfield.c:1058]: Can the `le` argument of `PyFloat_{Un}Pack*` be replaced with three-value `endianness` ( `big`, `little`, and `native`)? The `native` option would allow to move all `#ifdef WORDS_BIGENDIAN` fiddling inside the (up)pack functions.

File: Lib/test/test_float.py [hunk 3]
```diff
@@ -652,8 +659,9 @@ def test_float_specials_do_unpack(self):
             struct.unpack(fmt, data)
 
     @support.requires_IEEE_754
+    @unittest.skipIf(_testcapi is None, 'need _testcapi')
```
Comment (kumaraditya303) [inline Lib/test/test_float.py]:
```suggestion
    @unittest.skipIf(_testcapi is None, 'needs _testcapi')
```

File: Lib/test/test_float.py [hunk 4]
```diff
@@ -1488,5 +1496,69 @@ def __init__(self, value):
         self.assertEqual(getattr(f, 'foo', 'none'), 'bar')
 
 
+# Test PyFloat_Pack2(), PyFloat_Pack4() and PyFloat_Pack8()
+# Test PyFloat_Unpack2(), PyFloat_Unpack4() and PyFloat_Unpack8()
+BIG_ENDIAN = 0
+LITTLE_ENDIAN = 1
+ESPILON = {
```
Comment (kumaraditya303) [inline Lib/test/test_float.py]: Did you mean `epsilon` instead?

File: Lib/test/test_float.py [hunk 5]
```diff
@@ -1488,5 +1496,69 @@ def __init__(self, value):
...
+    2: 2.0 ** -11,  # binary16
+    4: 2.0 ** -24,  # binary32
+    8: 2.0 ** -53,  # binary64
+}
+
+@unittest.skipIf(_testcapi is None, 'need _testcapi')
```
Comment (kumaraditya303) [inline Lib/test/test_float.py]:
```suggestion
@unittest.skipIf(_testcapi is None, 'needs _testcapi')
```

File: Modules/_ctypes/cfield.c (hunk 2 above)
Comment (vstinner) [reply Modules/_ctypes/cfield.c:1058]: If this parameter is changed, it will be way harder to provide this function in pythoncapi_compat to older Python versions: https://github.com/pythoncapi/pythoncapi_compat/pull/26 And it maybe become more complicated to write a single code base supporting old and new Python versions.

It's easy to implement the "native" behavior by checking if your system uses little endian or not with your preferred way.

* marshal uses little endian
* struct supports big and little endian
* pickle uses big endian
* ctypes supports big and little endian
* msgpack uses big endian

So far, I didn't see a module using the "native endian". For a serialization function, it's not convenient.

File: Doc/c-api/float.rst [hunk 6]
```diff
@@ -76,3 +76,83 @@ Floating Point Objects
...
+
+
+Pack and Unpack functions
+=========================
+
+The :mod:`struct` and :mod:`pickle` (at least) modules need an efficient
+platform- independent way to store floating-point values as byte strings. The
+Pack routines produce a bytes string from a C :c:type:`double`, and the Unpack
+routines produce a C :c:type:`double` from such a bytes string. The suffix
+(4 or 8) specifies the number of bytes in the bytes string.
```
Comment (methane) [inline Doc/c-api/float.rst]: 2, 4, and 8?
Intentionally ignore 2?

**Comment by vstinner** [comment]
So far, tests pass on:

* OS: Windows, macOS, FreeBSD, Linux (Debian, Fedora, Gentoo, RHEL8, SLES)
* Arch: x86 (32 bit, x87?, little endian), x86-64 (64 bit, SSE?, little endian), ARM64 (little endian), PPC64 (big endian), s390x (big endian)

File: Doc/c-api/float.rst (hunk 6 above)
Comment (vstinner) [reply Doc/c-api/float.rst]: I copied the doc from the header file, I will add 2 once buildbot jobs complete.

**Review by methane — APPROVED** [review]

**Comment by vstinner** [comment]
Hum, I'm not sure if the PR is related to the USAN failure.

AMD64 Arch Linux Usan PR failed:

```
test_ints (ctypes.test.test_bitfields.C_Test) ... ok
/buildbot/buildarea/pull_request.pablogsal-arch-x86_64.clang-ubsan/build/Modules/_ctypes/cfield.c:554:5: runtime error: shift exponent 18446744073709551614 is too large for 16-bit type 'short'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /buildbot/buildarea/pull_request.pablogsal-arch-x86_64.clang-ubsan/build/Modules/_ctypes/cfield.c:554:5 in 
```

It's a test on the ``h`` format code:

```
#define LOW_BIT(x)  ((x) & 0xFFFF)
#define NUM_BITS(x) ((x) >> 16)

#define GET_BITFIELD(v, size)                                           \
    if (NUM_BITS(size)) {                                               \
        v <<= (sizeof(v)*8 - LOW_BIT(size) - NUM_BITS(size));           \
        v >>= (sizeof(v)*8 - NUM_BITS(size));                           \

static PyObject *
h_get(void *ptr, Py_ssize_t size)
{
    short val;
    memcpy(&val, ptr, sizeof(val));
    GET_BITFIELD(val, size); // <==== HERE
    return PyLong_FromLong((long)val);
}

static struct fielddesc formattable[] = {
    ...
    { 'h', h_set, h_get, NULL, h_set_sw, h_get_sw},
    ...
};
```

**Comment by vstinner** [comment]
> AMD64 Arch Linux Usan PR failed

Oh, I fact 3 tests failed:  test_ctypes, test_faulthandler, test_hashlib.

But it's unrelated to the PR, UBSan is also failing without this PR. I created: https://bugs.python.org/issue46913

File: Doc/c-api/float.rst (hunk 6 above)
Comment (vstinner) [reply Doc/c-api/float.rst]: Fixed

File: Doc/c-api/float.rst [hunk 7]
```diff
@@ -76,3 +76,83 @@ Floating Point Objects
 .. c:function:: double PyFloat_GetMin()
 
    Return the minimum normalized positive float *DBL_MIN* as C :c:type:`double`.
+
+
+Pack and Unpack functions
+=========================
+
+The :mod:`struct` and :mod:`pickle` (at least) modules need an efficient
+platform- independent way to store floating-point values as byte strings. The
```
Comment (mdickinson) [inline Doc/c-api/float.rst]:
```suggestion
platform-independent way to store floating-point values as byte strings. The
```

File: Doc/c-api/float.rst [hunk 8]
```diff
@@ -76,3 +76,83 @@ Floating Point Objects
 .. c:function:: double PyFloat_GetMin()
 
    Return the minimum normalized positive float *DBL_MIN* as C :c:type:`double`.
+
+
+Pack and Unpack functions
+=========================
+
+The :mod:`struct` and :mod:`pickle` (at least) modules need an efficient
```
Comment (mdickinson) [inline Doc/c-api/float.rst]: These mentions of specific modules made sense in the comments, but should probably be removed for the docs. I'd suggest removing them:

> The pack and unpack functions provide an efficient platform-independent way [...]

File: Doc/c-api/float.rst [hunk 9]
```diff
@@ -76,3 +76,83 @@ Floating Point Objects
...
+platform- independent way to store floating-point values as byte strings. The
+Pack routines produce a bytes string from a C :c:type:`double`, and the Unpack
+routines produce a C :c:type:`double` from such a bytes string. The suffix
+(2, 4 or 8) specifies the number of bytes in the bytes string.
+
+On platforms that appear to use IEEE-754 formats these functions work by
+copying bits. On other platforms, the formats the 4- byte format (32-bit) is
```
Comment (mdickinson) [inline Doc/c-api/float.rst]:
```suggestion
copying bits. On other platforms, the formats the 4-byte format (32-bit) is
```

File: Doc/c-api/float.rst [hunk 10]
```diff
@@ -76,3 +76,83 @@ Floating Point Objects
...
+identical to the IEEE-754 single precision format, and the 8-byte format
+(64-bit) to the IEEE-754 double precision format, although the packing of INFs
```
Comment (mdickinson) [inline Doc/c-api/float.rst]: The 2-byte format should be mentioned, too. (And yes, it's identical to the IEEE 754 "half-precision" binary16 format.)

File: Doc/c-api/float.rst [hunk 11]
```diff
@@ -76,3 +76,83 @@ Floating Point Objects
...
+
+The :mod:`struct` and :mod:`pickle` (at least) modules need an efficient
+platform- independent way to store floating-point values as byte strings. The
+Pack routines produce a bytes string from a C :c:type:`double`, and the Unpack
+routines produce a C :c:type:`double` from such a bytes string. The suffix
+(2, 4 or 8) specifies the number of bytes in the bytes string.
+
+On platforms that appear to use IEEE-754 formats these functions work by
+copying bits. On other platforms, the formats the 4- byte format (32-bit) is
+identical to the IEEE-754 single precision format, and the 8-byte format
```
Comment (mdickinson) [inline Doc/c-api/float.rst]: Nit: "IEEE 754", not "IEEE-754"

Comment (mdickinson) [reply Doc/c-api/float.rst]: To make this unambiguous, I'd suggest `the IEEE 754 binary32 "single precision" format`, and similarly for double precision. It's reasonably easy to _guess_ that "IEEE 754 single precision" means the binary32 format, but it's not totally unambiguous, and it would be better not to have to guess. (IEEE 754 specifies 4-byte decimal formats, too, and the standard itself never uses the terms "single precision" or "double precision".)

File: Doc/c-api/float.rst [hunk 12]
```diff
@@ -76,3 +76,83 @@ Floating Point Objects
...
+(64-bit) to the IEEE-754 double precision format, although the packing of INFs
+and NaNs (if such things exist on the platform) isn't handled correctly, and
+attempting to unpack a bytes string containing an IEEE INF or NaN will raise an
+exception.
+
+On non-IEEE platforms with more precision, or larger dynamic range, than IEEE
```
Comment (mdickinson) [inline Doc/c-api/float.rst:99]: Presumably once we're assuming IEEE 754 for CPython, all this text can be dropped?

File: Doc/c-api/float.rst [hunk 13]
```diff
@@ -76,3 +76,83 @@ Floating Point Objects
...
+754 supports, not all values can be packed; on non-IEEE platforms with less
+precision, or smaller dynamic range, not all values can be unpacked. What
+happens in such cases is partly accidental (alas).
+
+.. versionadded:: 3.11
+
+Pack functions
+--------------
+
+The pack routines write 2, 4 or 8 bytes, starting at *p*. *le* is a bool
```
Comment (mdickinson) [inline Doc/c-api/float.rst]: Maybe reword this: `le` is an `int`, not a `bool`, so it likely makes more sense to talk about zero and nonzero instead of true and false.

File: Doc/c-api/float.rst [hunk 14]
```diff
@@ -76,3 +76,83 @@ Floating Point Objects
...
+most likely :exc:`OverflowError`).
+
+There are two problems on non-IEEE platforms:
+
+* What this does is undefined if *x* is a NaN or infinity.
+* -0.0 and +0.0 produce the same bytes string.
+
+.. c:function:: int PyFloat_Pack2(double x, unsigned char *p, int le)
+
+   Pack a C double as 16-bit.
```
Comment (mdickinson) [inline Doc/c-api/float.rst]: Suggest "2 bytes" instead of 16 bits? It's probably worth re-mentioning the IEEE 754 binary16 format here, too. Similar comments apply to the next two entries.

File: Doc/c-api/float.rst [hunk 15]
```diff
@@ -76,3 +76,83 @@ Floating Point Objects
...
+
+.. c:function:: int PyFloat_Pack8(double x, unsigned char *p, int le)
+
+   Pack a C double as 64-bit.
+
+
+Unpack functions
+----------------
+
+The unpack routines read 2, 4 or 8 bytes, starting at *p*.  *le* is a bool
```
Comment (mdickinson) [inline Doc/c-api/float.rst]: Similarly, suggest replacing "bool" with "int" and replacing "true" and "false" with "nonzero" and "zero".

File: Doc/c-api/float.rst [hunk 16]
```diff
@@ -76,3 +76,83 @@ Floating Point Objects
...
+happens in such cases is partly accidental (alas).
+
+.. versionadded:: 3.11
+
+Pack functions
+--------------
+
+The pack routines write 2, 4 or 8 bytes, starting at *p*. *le* is a bool
+argument, true if you want the bytes string in little-endian format (exponent
+last, at ``p+1``, ``p+3`` or ``p+7``), false if you want big-endian format
```
Comment (mdickinson) [inline Doc/c-api/float.rst]: Nitpick: the exponent isn't actually at `p+7` in the 64-bit case, since it doesn't fit in a single byte. But it's probably clear enough what's meant.

**Review by mdickinson — COMMENTED** [review]
I've left a few comments, mostly on the docs.

File: Doc/c-api/float.rst (hunk 12 above)
Comment (vstinner) [reply Doc/c-api/float.rst:99]: My plan is to first fix the Python 3.11 "regression" by exposing these functions as public C API functions. And then deprecate or remove support for non-IEEE 754 platforms.

**Comment by vstinner** [comment]
@mdickinson: I addressed your comments. Since the documentation is quite long and complicated, I decided to remove it from Include/cpython/floatobject.h.

File: Doc/c-api/float.rst (hunk 12 above)
Comment (vstinner) [reply Doc/c-api/float.rst:99]: Require IEEE 754 support is tracked by https://bugs.python.org/issue46917

**Comment by vstinner** [comment]
Thanks everyone for your valuable reviews! I merged my PR.

There is likely room for enhancement on the doc, but it will be easier to read it once it will be rendered as HTML online at: https://docs.python.org/dev/c-api/float.html First, I only moved the existing doc from comments to the Python doc. This PR was a great opportunity to update the outdated doc and fix it!

Dropping support for non-IEEE 754 floating pointer formats is now tracked by https://bugs.python.org/issue46917

**Comment by vstinner** [comment]
Functions added to pythoncapi_compat: https://github.com/python/pythoncapi_compat/commit/6d2c96e073bc9dcc9fd090ba195ade42da7a7f18

File: Modules/_ctypes/cfield.c (hunk 2 above)
Comment (vstinner) [reply Modules/_ctypes/cfield.c:1058]: I wrote https://github.com/python/cpython/pull/31832 to suggest using PY_BIG_ENDIAN macro to use the native endian.
